lib/puppet/rest/route.rb

```diff
@@ -0,0 +1,55 @@
+module Puppet::Rest
+  class Route
+    attr_reader :api, :srv_service, :default_server, :default_port
+
+    attr_reader :server, :port
+
+    # Create a Route containing information for querying the given API,
+    # hosted at a server determined either by SRV service or by the
+    # fallback server on the fallback port.
+    def initialize(api:, srv_service:, default_server:, default_port:)
+      @api = api
+      @srv_service = srv_service
+      @default_server= default_server
+      @default_port = default_port
+    end
+
+    # Return the appropriate server and port for this route
+    # @return [String, Integer] the server and port to use for the request
+    def select_server_and_port
+      unless @server && @port
+        if Puppet.settings[:use_srv_records]
+          Puppet::Network::Resolver.each_srv_record(Puppet.settings[:srv_domain], srv_service) do |srv_server, srv_port|
```

Comment:
Not blocking for this PR, but there's some overlap in functionality with how we select the server and port based on SRV records in this class and in `Puppet::Network::Resolver`. It would be nice for performance reasons to pass a `Resolv::DNS` object into this class' initialize method, and then resolve SRV records and fallback server/ports as needed.

Reply:
I am already caching the results of the lookup for the duration of the client. This code is lifted more or less directly from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/request.rb#L187-L218, which appears to do this resolution per request, so I think this is already a performance win over what we have now. When you suggest passing  in a `Resolv::DNS` object, are you think to circumvent `Puppet::Network::Resolver` entirely?

Reply:
Oh, I didn't realize it was caching. In looking closer, I didn't realize the rest client assumes a single server/host and SRV service name for its lifetime. I was assuming we'd be able to use the pool of connections for both certificate and catalog endpoints, in the default case where the puppetserver is the CA. Maybe drop the SRV logic until we get the basic functionality and routing to `puppet-ca/v1` done?

Reply:
Sounds good. It might help me to talk a little more in depth about the history of the SRV stuff before then, assuming you know enough to give me some more context of how and why we use it.